sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,333 @@
+    this.outputTimersCoders = new HashMap<>();
+    this.perEndpointDataByteCount = new HashMap<>();
+    this.perEndpointTimersByteCount = new HashMap<>();
+    this.perEndpointElementCount = new HashMap<>();
+    this.outboundObserver = outboundObserver;
+    this.dataBuffer = new HashMap<>();
+    this.timersBuffer = new HashMap<>();
+    if (timeLimit > 0) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    } else {
+      this.flushFuture = null;
+    }
+    this.processBundleRequestIdSupplier = processBundleRequestIdSupplier;
+  }
+
+  /**
+   * Register the outbound data logical endpoint for the transform alongside it's output value
+   * coder.
+   */
+  public <T> FnDataReceiver<T> registerOutputDataLocation(String pTransformId, Coder<T> coder) {
+    outputDataCoders.put(pTransformId, coder);
+    if (timeLimit > 0) {
+      return data -> {
+        synchronized (flushLock) {
+          accept(
+              dataBuffer.computeIfAbsent(pTransformId, id -> ByteString.newOutput()),
+              (Coder<T>) outputDataCoders.get(pTransformId),
+              perEndpointDataByteCount.computeIfAbsent(pTransformId, id -> new LongAdder()),
+              data);
+        }
+      };
+    }
+    return data ->
+        accept(
+            dataBuffer.computeIfAbsent(pTransformId, id -> ByteString.newOutput()),
+            (Coder<T>) outputDataCoders.get(pTransformId),
+            perEndpointDataByteCount.computeIfAbsent(pTransformId, id -> new LongAdder()),
+            data);
+  }
+
+  /**
+   * Register the outbound timers logical endpoint for the transform alongside it's output value
+   * coder.
+   */
+  public <T> FnDataReceiver<T> registerOutputTimersLocation(
+      String pTransformId, String timerFamilyId, Coder<T> coder) {
+    KV<String, String> timerKey = KV.of(pTransformId, timerFamilyId);
+    outputTimersCoders.put(timerKey, coder);
+    if (timeLimit > 0) {
+      return timers -> {
+        synchronized (flushLock) {
+          accept(
+              timersBuffer.computeIfAbsent(timerKey, key -> ByteString.newOutput()),
+              (Coder<T>) outputTimersCoders.get(timerKey),
+              perEndpointTimersByteCount.computeIfAbsent(timerKey, id -> new LongAdder()),
+              timers);
+        }
+      };
+    }
+    return timers ->
+        accept(
+            timersBuffer.computeIfAbsent(timerKey, key -> ByteString.newOutput()),
+            (Coder<T>) outputTimersCoders.get(timerKey),
+            perEndpointTimersByteCount.computeIfAbsent(timerKey, id -> new LongAdder()),
+            timers);
+  }
+
+  // Convenience method for unit tests.
```

Comment:
Move this method into the unit test class like `registerOutputLocation(BeamFnDataOutboundAggregator agg, LogicalEndpoint endpoint, Coder<T> coder)`

Reply:
Done.

Reply:
Done.